File: src/docforge/app/frontend/src/api/snippets.ts
// ====== Code Summary ======
// TypeScript mirror of the granular collection config-SNIPPET REST contract + its typed client.
// Shapes copied verbatim from the backend's Pydantic models (app/backend/routers/snippets/models.py)
// — nothing invented. Distinct from api/transfers.ts's whole-collection `.dcexport` bundle: a
// snippet is SYNCHRONOUS and config-only (one slice — pipeline / search / schema), so both calls
// here are plain request/response, no polling.

import { apiFetch, jsonInit } from "./http";

const COLLECTIONS_BASE = "/api/v1/collections";

/** The three config slices a snippet can carry. */
export type SnippetKind = "pipeline" | "search" | "schema";

/** File extension the frontend saves a snippet under — deliberately distinct from `.dcexport`. */
export const SNIPPET_FILE_EXTENSION = ".dfsnippet";

/** A portable, secret-masked, versioned wrapper around one collection-config slice. */
export interface CollectionSnippet {
  kind: SnippetKind;
  format_version: number;
  docforge_version: string;
  /** The slice payload: the graph blob for `pipeline`/`search`, or `{fields: [...]}` for `schema`. */
  body: Record<string, unknown>;
}

export interface SnippetImportResult {
  collection_id: string;
  kind: SnippetKind;
  /** Whether applying this snippet flags a reindex requirement — always false for a search snippet. */
  needs_reindex: boolean;
}

/** Export one config slice of a collection as a versioned, secret-masked snippet — synchronous. */
export function getSnippet(collectionId: string, kind: SnippetKind): Promise<CollectionSnippet> {
  return apiFetch(`${COLLECTIONS_BASE}/${collectionId}/snippets/${kind}`);
}

/**
 * Apply an inbound config snippet onto an existing collection — synchronous, config-only.
 *
 * 422s when the snippet's `format_version` is unsupported or its `kind` doesn't match `kind`.
 */
export function applySnippet(collectionId: string, kind: SnippetKind, snippet: CollectionSnippet): Promise<SnippetImportResult> {
  return apiFetch(`${COLLECTIONS_BASE}/${collectionId}/snippets/${kind}`, jsonInit("POST", snippet));
}

File: src/docforge_sdk/docforge_sdk/models/snippets.py
# ====== Code Summary ======
# Request/response models for the granular collection config-SNIPPET endpoints, mirrored field-for-
# field from the DocForge backend (app/backend/routers/snippets/models.py). Unlike the whole-collection
# `.dcexport` bundle (async, worker-built, carries data), a snippet is a small, SYNCHRONOUS, config-only
# wrapper around ONE slice of a collection's configuration (its ingest pipeline blob, its search blob,
# or its metadata schema) — secret-masked and format-versioned. ``SNIPPET_FILE_EXTENSION`` (``.dfsnippet``)
# is deliberately distinct from ``.dcexport`` so the two are never confused on disk.

# ====== Standard Library Imports ======
from typing import Any, Literal

# ====== Third-Party Library Imports ======
from pydantic import BaseModel, Field

# File extension a snippet is saved under — distinct from the whole-collection `.dcexport` bundle.
SNIPPET_FILE_EXTENSION = ".dfsnippet"

# The three config slices a snippet can carry.
SnippetKind = Literal["pipeline", "search", "schema"]


class CollectionSnippet(BaseModel):
    """
    A portable, secret-masked, versioned wrapper around one collection-config slice.

    The ``body`` shape depends on ``kind``: for ``pipeline`` / ``search`` it is the graph blob (with
    every provider ``api_key`` masked); for ``schema`` it is ``{"fields": [FieldSpec, ...]}``.

    Attributes:
        kind (SnippetKind): Which config slice this snippet carries.
        format_version (int): The snippet format version (gated on import).
        docforge_version (str): The producing build's version (provenance, informational).
        body (dict[str, Any]): The slice payload (see the kind-specific shape above).
    """

    kind: SnippetKind = Field(description="Which config slice this snippet carries.")
    format_version: int = Field(description="The snippet format version (gated on import).")
    docforge_version: str = Field(description="Producing build version (provenance, informational).")
    body: dict[str, Any] = Field(description="The slice payload (blob dict, or {'fields': [...]}).")


class SnippetImportResult(BaseModel):
    """
    The outcome of applying a config snippet to a collection.

    Attributes:
        collection_id (str): The target collection's UUID.
        kind (SnippetKind): The config slice that was applied.
        needs_reindex (bool): Whether applying this snippet flags a reindex requirement (embed-space
            or searchable-schema change); always false for a search snippet.
    """

    collection_id: str = Field(description="The target collection's UUID.")
    kind: SnippetKind = Field(description="The config slice that was applied.")
    needs_reindex: bool = Field(
        description="Whether applying this snippet flags a reindex requirement (embed-space or "
        "searchable-schema change); always false for a search snippet."
    )


__all__ = ["SNIPPET_FILE_EXTENSION", "SnippetKind", "CollectionSnippet", "SnippetImportResult"]
